patches/tModLoader/Terraria/WorldGen.TML.cs

```diff
@@ -7,4 +19,80 @@ public partial class WorldGen
 	private static double _timePass = 0.0; // Used to account for more precise time rates.
 
 	internal static void ClearGenerationPasses() => _generator?._passes.Clear();
+
+	internal static List<Tuple<GenPass, ILContext.Manipulator>> genPassILEdits;
+	internal static List<Tuple<GenPass, GenPassDetour>> genPassDetours;
+
+	public delegate void GenPassDetour(WorldGenLegacyMethod orig, GenerationProgress progress, GameConfiguration configuration);
+
+	/// <summary>
+	/// IL edits a PassLegacy during world generation. Use in ModifyWorldGenerationTasks or ModifyHardmodeTasks.
+	/// </summary>
+	/// <param name="task">The task to be IL edited</param>
+	/// <param name="callback">The method that IL edits the pass</param>
+	// TODO: Add support for hardmode tasks
+	public static void ModifyTask(GenPass task, ILContext.Manipulator callback)
+	{
+		HookEndpointManager.Modify(GetGenPassMethod(task), callback);
+		genPassILEdits.Add(Tuple.Create(task, callback));
+	}
+
+	/// <summary>
+	/// Detours a PassLegacy during world generation. Use in ModifyWorldGenerationTasks or ModifyHardmodeTasks.
+	/// </summary>
+	/// <param name="task">The task to be detoured</param>
+	/// <param name="hookDelegate">The detour method</param>
+	// TODO: make detours work
+	public static void DetourTask(GenPass task, GenPassDetour hookDelegate)
+	{
+		HookEndpointManager.Add(GetGenPassMethod(task), hookDelegate);
+		genPassDetours.Add(Tuple.Create(task, hookDelegate));
+	}
+
+	internal static void RemoveGenPassILEditsAndDetours()
+	{
+		foreach (var ilEdit in genPassILEdits) {
+			HookEndpointManager.Unmodify(GetGenPassMethod(ilEdit.Item1), ilEdit.Item2);
+		}
+
+		foreach (var detour in genPassDetours) {
+			HookEndpointManager.Remove(GetGenPassMethod(detour.Item1), detour.Item2);
+		}
+	}
+
+	// TODO: tidy up
+	internal static MethodInfo GetGenPassMethod(GenPass task)
```

Comment:
Recommend this and `DetourTask` above take a `PassLegacy`. Let the compiler push the requirement onto the user so they're aware.